035-cruise-control-scaling-brokers.md

```diff
@@ -0,0 +1,113 @@
+* [`/remove_broker`](https://github.com/linkedin/cruise-control/wiki/REST-APIs#decommission-a-list-of-brokers-from-the-kafka-cluster): this endpoint allows to trigger a rebalancing operation by moving replicas out from the target brokers to remove, in order to leave them "empty" and not hosting any replicas. It can be used just before the scaling down, for getting a proposal (as dryrun) and running the actual rebalancing.
+
+In order to use the above two endpoints in the Strimzi cluster operator, they would be added to the [`CruiseControlApi`](https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApi.java) interface and developing the corresponding implementation.
+
+### KafkaRebalance modes
+
+The current usage of the `KafkaRebalance` custom resource allows to get an optimization proposal and start a full-rebalance with data movement across all the brokers in the cluster.
+At Cruise Control API level, it's reflected by the `KafkaRebalanceAssemblyOperator` using the `/proposals` endpoint for getting an optimization proposal first and then the `/rebalance` endpoint for starting the full-rebalance.
+
+In order to start a rebalancing on a cluster being scaled up or down, the proposal is about adding a `spec.mode` field in the `KafkaRebalance` custom resource with the following possible values:
+
+* `full-rebalance`: it starts a full-rebalance across all the brokers in the cluster as it works today. It is the default behavior if not specified (for backward compatibility).
+* `add-broker`: it starts a rebalancing by moving replicas out from the already existing brokers to the new added ones after a scale up.
+* `remove-broker`: it starts a rebalancing by moving replicas out from the target brokers to remove before a scale down.
+
+When the rebalancing is related to cluster being scaled up or down, the newly added brokers or the ones that are going to be removed have to be specified through a new `spec.brokers` field as an array.
+This needs come from how the Cruise Control API endpoints works. There is a specific query parameter which needs that list.
+
+A `KafkaRebalance` custom resource would look like this.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  mode: <value> # full-rebalance, add-broker or remove-broker. Assuming full-rebalance by default if the field is not specified.
+  brokers:
+    - 3
+    - 4
+  goals:
+    - RackAwareGoal
+    - ReplicaCapacityGoal
+```
+
+### Flow
+
+The proposal doesn't assume any automation to run the rebalancing when the cluster is scaled up or down.
```

Comment:
I think this is quite important point which probably should not come up only on line 73. I guess it answers at least one of my previous comments. TBH at least for me, this is a bit misleading since from the title and intro I assumed it takes the full automation into account.

Reply:
Are you suggesting to mention it in the introduction as well?

Reply:
Either in the introduction or make the title more clear. The title was what really set my expectation. The title suggests it happens on scaling the cluster. I read that as _the operator does it automatically when scaling the cluster_. If it would have been more something like _Add new rebalance types to Kafka Rebalance resource to help scling the cluster up or down_, it would have been more clear that this is actually not about automatic operations but just extending KafkaRebalance.